File: src/desktop/handlers/session.ts
import { DesktopCommandMap } from '../contracts';
import { DesktopRuntime } from '../runtime';

export async function handleStartSession(
    runtime: DesktopRuntime,
    payload: DesktopCommandMap['start_session']['request']
): Promise<DesktopCommandMap['start_session']['response']> {
    return runtime.startSession(payload);
}

export async function handlePauseSession(
    runtime: DesktopRuntime,
    _payload: DesktopCommandMap['pause_session']['request']
): Promise<DesktopCommandMap['pause_session']['response']> {
    return runtime.pauseSession();
}

export async function handleResumeSession(
    runtime: DesktopRuntime,
    _payload: DesktopCommandMap['resume_session']['request']
): Promise<DesktopCommandMap['resume_session']['response']> {
    return runtime.resumeSession();
}

export async function handleStopSession(
    runtime: DesktopRuntime,
    _payload: DesktopCommandMap['stop_session']['request']
): Promise<DesktopCommandMap['stop_session']['response']> {
    return runtime.stopSession();
}

export async function handleGetSessionState(
    runtime: DesktopRuntime,
    _payload: DesktopCommandMap['get_session_state']['request']
): Promise<DesktopCommandMap['get_session_state']['response']> {
    return runtime.getSessionState();
}

File: src-tauri/src/commands/session.rs
use crate::*;

#[tauri::command]
pub(crate) fn start_session(app: AppHandle, request: StartSessionRequest) -> Result<SessionSnapshot, String> {
    let mut payload = serde_json::to_value(request)
        .map_err(|error| format!("Failed to serialize the session runtime request: {error}"))?;
    if let Value::Object(ref mut object) = payload {
        object.insert(
            "token".to_string(),
            match resolve_effective_token(&app)? {
                Some(token) => Value::String(token),
                None => Value::Null,
            },
        );
    }

    send_sidecar_request(&app, "start_session", payload)
}

#[tauri::command]
pub(crate) fn pause_session(app: AppHandle) -> Result<Option<SessionSnapshot>, String> {
    send_sidecar_request(&app, "pause_session", json!({}))
}

#[tauri::command]
pub(crate) fn resume_session(app: AppHandle) -> Result<Option<SessionSnapshot>, String> {
    send_sidecar_request(&app, "resume_session", json!({}))
}

#[tauri::command]
pub(crate) fn stop_session(app: AppHandle) -> Result<Option<SessionSnapshot>, String> {
    send_sidecar_request(&app, "stop_session", json!({}))
}

#[tauri::command]
pub(crate) fn get_session_state(app: AppHandle) -> Result<Option<SessionSnapshot>, String> {
    send_sidecar_request(&app, "get_session_state", json!({}))
}

#[tauri::command]
pub(crate) fn load_logs(app: AppHandle, request: LoadLogsRequest) -> Result<LogLoadResult, String> {
    send_sidecar_request(&app, "load_logs", request)
}

#[tauri::command]
pub(crate) fn discard_resume_session(app: AppHandle) -> Result<SenderStateRecord, String> {
    send_sidecar_request(&app, "discard_resume_session", json!({}))
}
